### ISSUE TITLE
Optimize JIT

### ISSUE THREAD

**Issue by aerkiaga** [aerkiaga/hmars#17]
This is actually a very important point. Currently (as of the last measure), the classic algorithm is faster than JIT, which, in turn, is slower than pMARS. So for now hMARS uses classic as default (which is optimized and fast, especially after GCC 8.2). However, classic is reserved for debugging, as it has many useful hooks, deactivated at compile time if no debugger is needed. Also, it has the problem that most its parameters are hardcoded, but making them configurable would impact performance, which is not the case with JIT. Making the JIT simulator faster would allow making any necessary changes to clasic in order to build a solid and configurable debugging framework.
The first point to consider is turning all instruction callbacks into nested functions, eliminating unnecessary parameter loading and call overhead. Also, the main loop would be compiled into JIT code, which would improve code cache locality.

**Comment by aerkiaga** [comment]
Nesting implementation will require the new front-end nesting capabilities currently being worked on on libjit.

**Comment by aerkiaga** [comment]
Single-thread performance as of 9b9251539d718219713f0ca35336431b3969bbf0 (seconds, less is better):
```
                master  nesting
Classic         42.4    45
JIT             60.8    60.5
```
Nesting hasn't been implemented yet, but it can be seen that compiling the main loop does not increase performance by itself.

**Comment by aerkiaga** [comment]
Now another approach is being taken, where instead of compiling every instruction into its own function, all instructions are compiled into a jump-table/switch statement every time the main loop is compiled.

**Comment by aerkiaga** [comment]
It has been done, although now self-test code is much slower, since it must recompile everything on each round. Would be cool if LibJIT allowed to allocate/free functions separately.

**Comment by aerkiaga** [comment]
Fixed that in 527bbd14c2e8205e483b5111e37732f19b7e9416.

**Comment by aerkiaga** [comment]
Current performance (seconds, less is better):

```
hMARS    classic   JIT
master   41.3      61.0
jtable   42.7      53.4

pMARS    54.2
exMARS   28.6
```

Finally, the JIT version is faster than pMARS! More optimization would make even faster, but for now, this is enough.
